plugins/lidar.c

```diff
@@ -0,0 +1,365 @@
+#define AFE_DEVICE "ad5627"
+
+
+static struct iio_context *ctx;
+static struct iio_device *pulse_dev;
+static struct iio_device *afe_dev;
+static struct iio_channel *pulse_ch0;
+static struct iio_channel *afe_ch0;
+static struct iio_channel *afe_ch1;
+
+static struct iio_widget widgets[50];
+static unsigned int num_widgets = 0;
+
+static const gdouble khz_scale = 1000.0;
+static const gdouble inv_scale = -1.0;
+
+static int auto_cfg_prev_values[4];
+
+static GtkSpinButton *manual_ch_btns[4];
+static GtkSpinButton *auto_cfg_btns[4];
+
+static gint this_page;
+static gboolean plugin_detached;
+
+static void save_widget_value(GtkWidget *widget, struct iio_widget *iio_w)
+{
+	iio_w->save(iio_w);
+}
+
+static void make_widget_update_signal_based(struct iio_widget *widgets,
+	unsigned int num_widgets)
+{
+	char signal_name[25];
+	unsigned int i;
+
+	for (i = 0; i < num_widgets; i++) {
+		if (GTK_IS_CHECK_BUTTON(widgets[i].widget))
+			sprintf(signal_name, "%s", "toggled");
+		else if (GTK_IS_TOGGLE_BUTTON(widgets[i].widget))
+			sprintf(signal_name, "%s", "toggled");
+		else if (GTK_IS_SPIN_BUTTON(widgets[i].widget))
+			sprintf(signal_name, "%s", "value-changed");
+		else if (GTK_IS_COMBO_BOX_TEXT(widgets[i].widget))
+			sprintf(signal_name, "%s", "changed");
+		else
+			printf("unhandled widget type, attribute: %s\n", widgets[i].attr_name);
+
+		if (GTK_IS_SPIN_BUTTON(widgets[i].widget) &&
+			widgets[i].priv_progress != NULL) {
+				iio_spin_button_progress_activate(&widgets[i]);
+		} else {
+			g_signal_connect(G_OBJECT(widgets[i].widget), signal_name, G_CALLBACK(save_widget_value), &widgets[i]);
+		}
+	}
+}
+
+static void manual_ch_set_cb(void)
+{
+	const int len = 9; // 4 digits, 4 spaces and NUL character
+	char ch_config[len];
+
+	snprintf(ch_config, len, "%i %i %i %i",
+		gtk_spin_button_get_value_as_int(manual_ch_btns[0]),
+		gtk_spin_button_get_value_as_int(manual_ch_btns[1]),
+		gtk_spin_button_get_value_as_int(manual_ch_btns[2]),
+		gtk_spin_button_get_value_as_int(manual_ch_btns[3]));
+
+	iio_device_attr_write_raw(pulse_dev,
+		"sequencer_manual_chsel", ch_config, len);
+}
+
+static void auto_cfg_set_cb(void)
+{
+	const int len = 9; // 4 digits, 4 spaces and NUL character
+	char ch_config[len];
+
+	snprintf(ch_config, len, "%i %i %i %i",
+		gtk_spin_button_get_value_as_int(auto_cfg_btns[0]),
+		gtk_spin_button_get_value_as_int(auto_cfg_btns[1]),
+		gtk_spin_button_get_value_as_int(auto_cfg_btns[2]),
+		gtk_spin_button_get_value_as_int(auto_cfg_btns[3]));
+
+	iio_device_attr_write_raw(pulse_dev,
+		"sequencer_auto_cfg", ch_config, len);
+}
+
+static void auto_cfg_button_changed_cb(GtkSpinButton *btn)
+{
+	// Force the buttons value to have a valid permutation of (0, 1, 2, 3)
+	int i, j, idx;
+	int old_val;
+	int crt_val = gtk_spin_button_get_value_as_int(btn);
+
+	// Find the index for the button that this callback has been called (current button)
+	for (i = 0; i < 4; i++) {
+		if (btn == auto_cfg_btns[i])
+			break;
+	}
+
```

Comment:
oh ; wait;
codacy found something: `Array 'auto_cfg_prev_values[4]' accessed at index 4, which is out of bounds.`

maybe check if `i == 4` here;